bin/cheribsdtest/cheribsdtest_lazy_bind.c

```diff
@@ -0,0 +1,47 @@
+/*-
+ * Copyright (c) 2021 Jessica Clarke
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/types.h>
+
+#include <cheri/cheri.h>
+#include <cheri/cheric.h>
+
+#include <err.h>
+
+#include <cheribsdtest_dynamic.h>
+
+#include "cheribsdtest.h"
+
+CHERIBSDTEST(test_lazy_bind_args,
+    "Check that lazy binding preserves capability argument metadata")
+{
+	void * __capability cap, * __capability cap2;
+
+	cap = (__cheri_tocap void * __capability)&cap;
+	cap2 = cheribsdtest_dynamic_identity_cap(cap);
```

Comment:
Maybe add a short comment here stating that cheribsdtest_dynamic_identity_cap() is defined in another shared library and calling it will go via the lazy binding resolver. MIght also want to check that LD_BIND_NOW is not set?

Reply:
And also the flag in in _DYNAMIC I guess.

Reply:
Even then it's up to rtld and the toolchain whether it wants to support lazy binding, it's perfectly valid to not do it. Currently we run this test for -purecap and all of those things will look like it's supported but still lazy binding isn't in use because it's not implemented. I'm not sure it's worth the trouble.